util/testutil/workers/dockerd.go

```diff
@@ -18,49 +19,92 @@ import (
+		FeatureCacheImport,
+		FeatureCacheBackendAzblob,
+		FeatureCacheBackendGha,
+		FeatureCacheBackendLocal,
+		FeatureCacheBackendRegistry,
+		FeatureCacheBackendS3,
+		FeatureDirectPush,
+		FeatureImageExporter,
+		FeatureMultiCacheExport,
+		FeatureMultiPlatform,
+		FeatureOCIExporter,
+		FeatureOCILayout,
+		FeatureProvenance,
+		FeatureSBOM,
+		FeatureSecurityMode,
+		FeatureCNINetwork,
+	}
+	containerdUnsupportedFeatures = []string{
+		FeatureSecurityMode,
+		FeatureCNINetwork,
+	}
+)
+
 // InitDockerdWorker registers a dockerd worker with the global registry.
 func InitDockerdWorker() {
 	integration.Register(&Moby{
-		ID:         "dockerd",
-		IsRootless: false,
-		Unsupported: []string{
-			FeatureCacheExport,
-			FeatureCacheImport,
-			FeatureCacheBackendAzblob,
-			FeatureCacheBackendGha,
-			FeatureCacheBackendLocal,
-			FeatureCacheBackendRegistry,
-			FeatureCacheBackendS3,
-			FeatureDirectPush,
-			FeatureImageExporter,
-			FeatureMultiCacheExport,
-			FeatureMultiPlatform,
-			FeatureOCIExporter,
-			FeatureOCILayout,
-			FeatureProvenance,
-			FeatureSBOM,
-			FeatureSecurityMode,
-			FeatureCNINetwork,
-		},
+		ID:          "dockerd",
+		IsRootless:  false,
+		Unsupported: mobyUnsupportedFeatures,
 	})
 	integration.Register(&Moby{
 		ID:                    "dockerd-containerd",
 		IsRootless:            false,
 		ContainerdSnapshotter: true,
-		Unsupported: []string{
-			FeatureSecurityMode,
-			FeatureCNINetwork,
-		},
+		Unsupported:           containerdUnsupportedFeatures,
 	})
+
+	// e.g. `dockerd-26.0=/opt/dockerd-26.0/bin,dockerd-25.0=/opt/dockerd-25.0/bin`
+	if s := os.Getenv("BUILDKIT_INTEGRATION_DOCKERD_EXTRA"); s != "" {
+		entries := strings.Split(s, ",")
+		for _, entry := range entries {
+			p1 := strings.Split(strings.TrimSpace(entry), "=")
+			if len(p1) != 2 {
+				panic(errors.Errorf("unexpected BUILDKIT_INTEGRATION_DOCKERD_EXTRA: %q", s))
+			}
+			fullname, bin := p1[0], p1[1]
+			if _, err := os.Stat(bin); err != nil {
+				panic(errors.Wrapf(err, "unexpected BUILDKIT_INTEGRATION_DOCKERD_EXTRA: %q", s))
+			}
+			p2 := strings.Split(strings.TrimSpace(fullname), "-")
+			if len(p2) != 2 {
+				panic(errors.Errorf("unexpected BUILDKIT_INTEGRATION_DOCKERD_EXTRA: %q", s))
+			}
+			_, ver := p2[0], p2[1]
+			if ver == "" {
+				panic(errors.Errorf("unexpected BUILDKIT_INTEGRATION_DOCKERD_EXTRA: %q", s))
+			}
+			integration.Register(&Moby{
+				ID:          fmt.Sprintf("dockerd-%s", ver),
+				Binary:      filepath.Join(bin, "dockerd"),
+				IsRootless:  false,
+				ExtraEnv:    []string{fmt.Sprintf("PATH=%s:%s", bin, os.Getenv("PATH"))},
+				Unsupported: mobyUnsupportedFeatures,
+			})
+			integration.Register(&Moby{
+				ID:                    fmt.Sprintf("dockerd-containerd-%s", ver),
+				Binary:                filepath.Join(bin, "dockerd"),
+				IsRootless:            false,
+				ContainerdSnapshotter: true,
+				ExtraEnv:              []string{fmt.Sprintf("PATH=%s:%s", bin, os.Getenv("PATH"))},
```

Comment:
If we only use `ExtraEnv` for the `PATH`, why isn't just overriding the `Binary` sufficient?

Reply:
I want that for https://github.com/docker/buildx/pull/2307 so I can get extra env set from backend to run the right bins for `docker` client: https://github.com/docker/buildx/pull/2307/files#diff-6db6da28b382b90cebe937a99fedeb6c8927535c220fd1c9a40f644d4ae90ee7R58-R60